### PR TITLE
Add `MethodFilter::CONNECT`

### PR DESCRIPTION
**Pull request by SabrinaJewson**
This makes it easiër to support [HTTP/2 Extended Connect](https://www.rfc-editor.org/rfc/rfc8441.html#section-4) with Axum.

Suggested originally in #2894. I didn’t modify the documentation of WebSockets, because `any(handler)` looks a lot nicer than `on(MethodFilter::GET.or(MethodFilter::CONNECT), handler)` and there aren’t many disadvantages to using `any` (incorrect methods are rejected anyway).

### CODE REVIEW

File: axum-extra/src/routing/mod.rs [hunk 1]
```diff
@@ -131,6 +131,19 @@ pub trait RouterExt<S>: sealed::Sealed {
...
         P: TypedPath;
 
+    /// Add a typed `CONNECT` route to the router.
+    ///
+    /// The path will be inferred from the first argument to the handler function which must
+    /// implement [`TypedPath`].
+    ///
+    /// See [`TypedPath`] for more details and examples.
+    #[cfg(feature = "typed-routing")]
+    fn typed_connect<H, T, P>(self, handler: H) -> Self
```
Comment (jplatte) [inline axum-extra/src/routing/mod.rs:141]: Would be nice to have a changelog entry for this too.

File: axum/src/routing/method_filter.rs [hunk 2]
```diff
@@ -9,6 +9,24 @@ use std::{
 pub struct MethodFilter(u16);
 
 impl MethodFilter {
+    /// Match `CONNECT` requests.
+    ///
+    /// This is useful for implementing HTTP/2’s [extended CONNECT method],
```
Comment (jplatte) [inline axum/src/routing/method_filter.rs:14]: Weird nitpick, but I think it's pretty uncommon to use that unicode `’` in source code, even comments? I wouldn't even know how to type it 😅

File: axum/src/routing/method_filter.rs [hunk 3]
```diff
@@ -9,6 +9,24 @@ use std::{
...
+    /// in which the `:protocol` pseudoheader is read
+    /// (using [`hyper::ext::Protocol`])
+    /// and the connection upgraded to a bidirectional byte stream
+    /// (using [`hyper::upgrade::on`]).
+    ///
+    /// As seen in the [HTTP Upgrade Token Registry],
+    /// common uses include WebSockets and proxying UDP or IP –
+    /// though note that when using [`WebSocketUpgrade`]
+    /// it’s more useful to use [`any`](crate::routing::any)
```
Comment (jplatte) [inline axum/src/routing/method_filter.rs:23]: Same here.

File: axum/CHANGELOG.md [hunk 4]
```diff
@@ -15,8 +15,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - **breaking:** Upgrade matchit to 0.8, changing the path parameter syntax from `/:single` and `/*many`
   to `/{single}` and `/{*many}`; the old syntax produces a panic to avoid silent change in behavior ([#2645])
 - **change:** Update minimum rust version to 1.75 ([#2943])
-- **added:** Add support WebSockets over HTTP/2.
+- **added:** Add support for WebSockets over HTTP/2.
   They can be enabled by changing `get(ws_endpoint)` handlers to `any(ws_endpoint)`.
+- **added:** Add `MethodFilter::CONNECT`.
```
Comment (jplatte) [inline axum/CHANGELOG.md:20]:
```suggestion
- **added:** Add `MethodFilter::CONNECT`, `routing::connect[_service]` and `MethodRouter::connect[_service]` ([#2961])
```

plus link reference to `[#2961]`

**Review by jplatte — COMMENTED** [review]
Thanks!

File: axum/src/routing/method_filter.rs (hunk 2 above)
Comment (SabrinaJewson) [reply axum/src/routing/method_filter.rs]: Hah, that’s force of habit from me 😅. Fixed now.

**Comment by SabrinaJewson** [comment]
I reälized the WebSockets PR was in the wrong section in the changelog, so I fixed that too.

File: axum-extra/CHANGELOG.md [hunk 5]
```diff
@@ -5,6 +5,12 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog],
 and this project adheres to [Semantic Versioning].
 
+# Unreleased
+
+- **added:** Add `RouterExt::typed_connect` ([#2961]).
```
Comment (jplatte) [inline axum-extra/CHANGELOG.md:10]:
```suggestion
- **added:** Add `RouterExt::typed_connect` ([#2961])
```

File: axum/CHANGELOG.md [hunk 6]
```diff
@@ -5,6 +5,16 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+# Unreleased
+
+- **added:** Add support for WebSockets over HTTP/2 ([#2894]).
+  They can be enabled by changing `get(ws_endpoint)` handlers to `any(ws_endpoint)`.
+- **added:** Add `MethodFilter::CONNECT`, `routing::connect[_service]`
+  and `MethodRouter::connect[_service]` ([#2961]).
```
Comment (jplatte) [inline axum/CHANGELOG.md:13]:
```suggestion
- **added:** Add support for WebSockets over HTTP/2 ([#2894]).
  They can be enabled by changing `get(ws_endpoint)` handlers to `any(ws_endpoint)`
- **added:** Add `MethodFilter::CONNECT`, `routing::connect[_service]`
  and `MethodRouter::connect[_service]` ([#2961])
```

**Review by jplatte — APPROVED** [review]

**Pull request merged**